### ISSUE TITLE
Convenient way to create new instances

### ISSUE THREAD

**Issue by hastebrot** [ianstormtaylor/superstruct#382]
Is there a more convenient way to validate objects on creation than using `coerce(data, struct)`? I've introduced `withNew()` to augment my struct with `new()`.

~~~js
const node = Node.new({ key: 1, left: Node.new({ key: 2 }) })
console.log(S.is(node, Node))
~~~

~~~js
const withNew = struct => ({ 
  ...struct, 
  new(data) { 
    return S.coerce(data, struct) 
  }
})

const Node = withNew(S.type({
  key: S.union([S.number(), S.string()]),
  rank: S.optional(S.number()),
  left: S.optional(S.lazy(() => Node)),
  right: S.optional(S.lazy(() => Node)),
}))
~~~

Code example: https://observablehq.com/@hastebrot/hello-superstruct@39

`superstructVersion = "0.10.11"`

---

**Edit:** Pinned ObservableHQ notebook to a version that uses `withNew()`.

**Comment by mikestopcontinues** [comment]
This isn't much cleaner, but it saves you having to wrap all of your type definitions...

```js
const factoryType = (...args) => {
  const struct = S.type(...args);
  return {...struct, new: (data) => S.coerce(data, struct)}
};

const Node = factoryType({key, rank, left, right});
Node.new({left, right});
```

You can make it more constructor-esque:

```js
const factoryType = (...args) => {
  const struct = S.type(...args);

  const factory = (data) => S.coerce(data, struct);
  factory.validate = (data) => S.validate(data, struct);
  return factory;
};

const Node = factoryType({key, rank, left, right});
const someNode = Node({left, right});
Node.validate(someNode);
```

**Comment by ianstormtaylor** [comment]
This is solved in https://github.com/ianstormtaylor/superstruct/pull/383 with `assign`.

**Comment by ianstormtaylor** [comment]
This is solved in #383 by renaming `coerce` to `create`, which is available as `Struct.create` as well.
